File: ring-android/app/src/main/java/cx/ring/dependencyinjection/ServiceInjectionModule.java
/*
 *  Copyright (C) 2004-2021 Savoir-faire Linux Inc.
 *
 *  Author: Thibault Wittemberg <thibault.wittemberg@savoirfairelinux.com>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */
package cx.ring.dependencyinjection;

import android.content.Context;

import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;

import javax.inject.Named;
import javax.inject.Singleton;

import cx.ring.application.JamiApplication;
import net.jami.facades.ConversationFacade;
import net.jami.services.AccountService;
import net.jami.services.CallService;
import net.jami.services.ContactService;
import cx.ring.services.ContactServiceImpl;
import net.jami.services.DaemonService;
import net.jami.services.DeviceRuntimeService;
import cx.ring.services.DeviceRuntimeServiceImpl;
import net.jami.services.HardwareService;
import cx.ring.services.HardwareServiceImpl;
import net.jami.services.HistoryService;
import cx.ring.services.HistoryServiceImpl;
import net.jami.services.LogService;
import cx.ring.services.LogServiceImpl;
import net.jami.services.NotificationService;
import cx.ring.services.NotificationServiceImpl;
import net.jami.services.PreferencesService;
import cx.ring.services.SharedPreferencesServiceImpl;
import net.jami.services.VCardService;
import cx.ring.services.VCardServiceImpl;
import net.jami.utils.Log;
import dagger.Module;
import dagger.Provides;
import io.reactivex.Scheduler;
import io.reactivex.android.schedulers.AndroidSchedulers;

@Module
public class ServiceInjectionModule {

    private final JamiApplication mJamiApplication;

    public ServiceInjectionModule(JamiApplication app) {
        mJamiApplication = app;
    }

    @Provides
    @Singleton
    PreferencesService provideSettingsService() {
        SharedPreferencesServiceImpl settingsService = new SharedPreferencesServiceImpl();
        mJamiApplication.getInjectionComponent().inject(settingsService);
        return settingsService;
    }

    @Provides
    @Singleton
    HistoryService provideHistoryService() {
        HistoryServiceImpl historyService = new HistoryServiceImpl();
        mJamiApplication.getInjectionComponent().inject(historyService);
        return historyService;
    }

    @Provides
    @Singleton
    LogService provideLogService() {
        LogService service = new LogServiceImpl();
        Log.injectLogService(service);
        return service;
    }

    @Provides
    @Singleton
    NotificationService provideNotificationService() {
        NotificationServiceImpl service = new NotificationServiceImpl();
        mJamiApplication.getInjectionComponent().inject(service);
        service.initHelper();
        return service;
    }

    @Provides
    @Singleton
    DeviceRuntimeService provideDeviceRuntimeService(LogService logService) {
        DeviceRuntimeServiceImpl runtimeService = new DeviceRuntimeServiceImpl();
        mJamiApplication.getInjectionComponent().inject(runtimeService);
        runtimeService.loadNativeLibrary();
        return runtimeService;
    }

    @Provides
    @Singleton
    DaemonService provideDaemonService(DeviceRuntimeService deviceRuntimeService) {
        DaemonService daemonService = new DaemonService(deviceRuntimeService);
        mJamiApplication.getInjectionComponent().inject(daemonService);
        return daemonService;
    }

    @Provides
    @Singleton
    CallService provideCallService() {
        CallService callService = new CallService();
        mJamiApplication.getInjectionComponent().inject(callService);
        return callService;
    }

    @Provides
    @Singleton
    AccountService provideAccountService() {
        AccountService accountService = new AccountService();
        mJamiApplication.getInjectionComponent().inject(accountService);
        return accountService;
    }

    @Provides
    @Singleton
    HardwareService provideHardwareService(Context context) {
        HardwareServiceImpl hardwareService = new HardwareServiceImpl(context);
        mJamiApplication.getInjectionComponent().inject(hardwareService);
        return hardwareService;
    }

    @Provides
    @Singleton
    ContactService provideContactService(PreferencesService sharedPreferencesService) {
        ContactServiceImpl contactService = new ContactServiceImpl();
        mJamiApplication.getInjectionComponent().inject(contactService);
        return contactService;
    }

    @Provides
    @Singleton
    ConversationFacade provideConversationFacade(
            HistoryService historyService,
            CallService callService,
            ContactService contactService,
            AccountService accountService,
            NotificationService notificationService) {
        ConversationFacade conversationFacade = new ConversationFacade(historyService, callService, accountService, contactService, notificationService);
        mJamiApplication.getInjectionComponent().inject(conversationFacade);
        return conversationFacade;
    }

    @Provides
    @Singleton
    VCardService provideVCardService(Context context) {
        return new VCardServiceImpl(context);
    }

    @Provides
    @Named("DaemonExecutor")
    @Singleton
    ScheduledExecutorService provideDaemonExecutorService() {
        return Executors.newSingleThreadScheduledExecutor(r -> new Thread(r, "DRing"));
    }

    @Provides
    @Named("UiScheduler")
    @Singleton
    Scheduler provideUiScheduler() {
        return AndroidSchedulers.mainThread();
    }
}

File: jami-android/app/src/main/java/cx/ring/dependencyinjection/ServiceInjectionModule.kt
/*
 *  Copyright (C) 2004-2025 Savoir-faire Linux Inc.
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program. If not, see <https://www.gnu.org/licenses/>.
 */
package cx.ring.dependencyinjection

import android.content.Context
import cx.ring.services.*
import cx.ring.utils.DeviceUtils
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.android.qualifiers.ApplicationContext
import dagger.hilt.components.SingletonComponent
import io.reactivex.rxjava3.core.Scheduler
import net.jami.services.ConversationFacade
import net.jami.services.*
import net.jami.utils.Log
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import javax.inject.Named
import javax.inject.Singleton

@Module
@InstallIn(SingletonComponent::class)
object ServiceInjectionModule {
    @Provides
    @Singleton
    fun provideSettingsService(@ApplicationContext appContext: Context, accountService: AccountService, deviceService: DeviceRuntimeService): PreferencesService {
        return SharedPreferencesServiceImpl(appContext, accountService, deviceService)
    }

    @Provides
    @Singleton
    fun provideHistoryService(@ApplicationContext appContext: Context): HistoryService {
        return HistoryServiceImpl(appContext)
    }

    @Provides
    @Singleton
    fun provideLogService(): LogService {
        val service: LogService = LogServiceImpl()
        Log.injectLogService(service)
        return service
    }

    @Provides
    @Singleton
    fun provideNotificationService(@ApplicationContext appContext: Context, accountService: AccountService,
                                   contactService: ContactService,
                                   preferencesService: PreferencesService,
                                   deviceRuntimeService: DeviceRuntimeService,
                                   callService: CallService): NotificationService {
        return NotificationServiceImpl(appContext, accountService, contactService, preferencesService, deviceRuntimeService, callService)
    }

    @Provides
    @Singleton
    fun provideDeviceRuntimeService(
        @ApplicationContext appContext: Context, @Named("DaemonExecutor") executor: ScheduledExecutorService, logService: LogService
    ): DeviceRuntimeService {
        val runtimeService = DeviceRuntimeServiceImpl(appContext, executor, logService)
        runtimeService.loadNativeLibrary()
        return runtimeService
    }

    @Provides
    @Singleton
    fun provideDaemonService(deviceRuntimeService: DeviceRuntimeService,
                             @Named("DaemonExecutor") executor: ScheduledExecutorService,
                             callService: CallService,
                             hardwareService: HardwareService,
                             accountService: AccountService): DaemonService {
        return DaemonService(deviceRuntimeService, executor, callService, hardwareService, accountService)
    }

    @Provides
    @Singleton
    fun provideCallService(@ApplicationContext appContext: Context,
                           @Named("DaemonExecutor") executor : ScheduledExecutorService,
                           contactService: ContactService,
                           accountService: AccountService,
                           deviceRuntimeService: DeviceRuntimeService): CallService {
        return CallServiceImpl(appContext, executor, contactService, accountService, deviceRuntimeService)
    }

    @Provides
    @Singleton
    fun provideAccountService(@Named("DaemonExecutor") executor : ScheduledExecutorService,
                              historyService : HistoryService,
                              deviceRuntimeService : DeviceRuntimeService,
                              vCardService : VCardService): AccountService {
        return AccountService(executor, historyService, deviceRuntimeService, vCardService)
    }

    @Provides
    @Singleton
    fun provideHardwareService(@ApplicationContext appContext: Context,
                               @Named("DaemonExecutor") executor : ScheduledExecutorService,
                               preferenceService: PreferencesService,
                               @Named("UiScheduler") uiScheduler: Scheduler): HardwareService {
        return HardwareServiceImpl(appContext, executor, preferenceService, uiScheduler)
    }

    @Provides
    @Singleton
    fun provideContactService(@ApplicationContext appContext: Context,
                              preferenceService: PreferencesService,
                              accountService: AccountService): ContactService {
        return ContactServiceImpl(appContext, preferenceService, accountService)
    }

    @Provides
    @Singleton
    fun provideConversationFacade(
        historyService: HistoryService,
        callService: CallService,
        contactService: ContactService,
        accountService: AccountService,
        notificationService: NotificationService,
        hardwareService: HardwareService,
        deviceRuntimeService: DeviceRuntimeService,
        preferencesService: PreferencesService
    ): ConversationFacade {
        return ConversationFacade(
            historyService,
            callService,
            accountService,
            contactService,
            notificationService,
            hardwareService,
            deviceRuntimeService,
            preferencesService
        )
    }

    @Provides
    @Singleton
    fun provideVCardService(@ApplicationContext appContext: Context): VCardService {
        return VCardServiceImpl(appContext)
    }

    @Provides
    @Named("DaemonExecutor")
    @Singleton
    fun provideDaemonExecutorService(): ScheduledExecutorService {
        return Executors.newSingleThreadScheduledExecutor { r: Runnable? -> Thread(r, "DRing") }
    }

    @Provides
    @Named("UiScheduler")
    @Singleton
    fun provideUiScheduler(): Scheduler {
        return DeviceUtils.uiScheduler
    }
}